crates/core/src/delta_datafusion/mod.rs

```diff
@@ -111,6 +118,150 @@ impl From<DataFusionError> for DeltaTableError {
+
+    fn input_schema(&self) -> DeltaResult<ArrowSchemaRef> {
+        _arrow_schema(self, false)
+    }
+}
+
+impl DataFusionMixins for EagerSnapshot {
+    fn arrow_schema(&self) -> DeltaResult<ArrowSchemaRef> {
+        self.snapshot().arrow_schema()
+    }
+
+    fn parse_predicate_expression(
+        &self,
+        expr: impl AsRef<str>,
+        df_state: &SessionState,
+    ) -> DeltaResult<Expr> {
+        self.snapshot().parse_predicate_expression(expr, df_state)
+    }
+
+    fn input_schema(&self) -> DeltaResult<ArrowSchemaRef> {
+        self.snapshot().input_schema()
+    }
+}
+
+impl DataFusionMixins for DeltaTableState {
+    fn arrow_schema(&self) -> DeltaResult<ArrowSchemaRef> {
+        self.snapshot.arrow_schema()
+    }
+
+    fn input_schema(&self) -> DeltaResult<ArrowSchemaRef> {
+        self.snapshot.input_schema()
+    }
+
+    fn parse_predicate_expression(
+        &self,
+        expr: impl AsRef<str>,
+        df_state: &SessionState,
+    ) -> DeltaResult<Expr> {
+        self.snapshot.parse_predicate_expression(expr, df_state)
+    }
+}
+
+fn _arrow_schema(snapshot: &Snapshot, wrap_partitions: bool) -> DeltaResult<ArrowSchemaRef> {
+    let meta = snapshot.metadata();
+    let fields = meta
+        .schema()?
+        .fields()
+        .iter()
+        .filter(|f| !meta.partition_columns.contains(&f.name().to_string()))
+        .map(|f| f.try_into())
+        .chain(
+            meta.schema()?
+                .fields()
+                .iter()
+                .filter(|f| meta.partition_columns.contains(&f.name().to_string()))
+                .map(|f| {
+                    let field = Field::try_from(f)?;
+                    let corrected = if wrap_partitions {
+                        match field.data_type() {
+                            // Only dictionary-encode types that may be large
+                            // // https://github.com/apache/arrow-datafusion/pull/5545
+                            DataType::Utf8
+                            | DataType::LargeUtf8
+                            | DataType::Binary
+                            | DataType::LargeBinary => {
+                                wrap_partition_type_in_dict(field.data_type().clone())
+                            }
+                            _ => field.data_type().clone(),
+                        }
+                    } else {
+                        field.data_type().clone()
+                    };
+                    Ok(field.with_data_type(corrected))
+                }),
+        )
+        .collect::<Result<Vec<Field>, _>>()?;
+
+    Ok(Arc::new(ArrowSchema::new(fields)))
+}
+
+pub(crate) trait DataFusionFileMixins {
+    /// Iterate over all files in the log matching a predicate
+    fn files_matching_predicate<'a>(
+        &'a self,
+        filters: &'a [Expr],
+    ) -> DeltaResult<impl Iterator<Item = Add>>;
+}
+
+pub(crate) fn files_matching_predicate<'a>(
+    snapshot: &'a EagerSnapshot,
+    filters: &[Expr],
```

Comment:
Does the missing lifetime matter here?

Reply:
I think the compiler is able to infer the lifetime of filters must be greater than or equal to `'a`. Had to introduce the lifetimes in the early stages to satisfy the compiler but I can try removing them now.